### ISSUE TITLE
about specific loss function

### ISSUE THREAD

**Issue by BXuan694** [pjreddie/darknet#1656]
In line 183 of yolo_layer.c, there is:
`l.delta[obj_index] = 1 - l.output[obj_index];
`
I think it is to compute the gradient of the square loss, but in some implementation, there can be:
`l.delta[obj_index] = l.output[obj_index] - 1;
`
what's the difference? Why does the second work?

**Comment by auguster** [comment]
I think it comes from the "square loss" where (a-b)^2 = (b-a)^2

**Comment by AlexeyAB** [comment]
> I think it is to compute the gradient of the square loss

It comes from Derivative of the binary cross-entropy loss function for the logistic function: https://peterroelants.github.io/posts/cross-entropy-logistic/#Derivative-of-the-cross-entropy-loss-function-for-the-logistic-function


> but in some implementation, there can be:
> l.delta[obj_index] = l.output[obj_index] - 1; 

Where did you see this?

**Comment by BXuan694** [comment]
> > I think it is to compute the gradient of the square loss
> 
> It comes from Derivative of the binary cross-entropy loss function for the logistic function: https://peterroelants.github.io/posts/cross-entropy-logistic/#Derivative-of-the-cross-entropy-loss-function-for-the-logistic-function
> 
> > but in some implementation, there can be:
> > l.delta[obj_index] = l.output[obj_index] - 1;
> 
> Where did you see this?

It should be so in Caffe, because negative gradient should pass in darknet. However,  positive gradient should pass in Caffe.

See:

Net::Update() in https://github.com/BVLC/caffe/blob/master/src/caffe/net.cpp
Blob<Dtype>::Update() in https://github.com/BVLC/caffe/blob/master/src/caffe/blob.cpp
void caffe_axpy() in https://github.com/BVLC/caffe/blob/master/src/caffe/util/math_functions.cpp

and

void update_convolutional_layer() in https://github.com/pjreddie/darknet/blob/f6d861736038da22c9eb0739dca84003c5a5e275/src/convolutional_layer.c
